prevent from publishing dimensions change event when app changes state

lbaldy opened the pull request:
## Summary

An issue raised in https://github.com/Expensify/App/issues/2727 has to be fixed in react native core. This impacts IOS only.

## Changelog

a) The body of the \_interfaceOrientationDidChange:

Here we make sure we run this function when the app is in active and also it was switching to fullscreeen or is not running in fullscreen.

```
- (void)_interfaceOrientationDidChange
{
  UIInterfaceOrientation nextOrientation = [RCTSharedApplication() statusBarOrientation];
  UIApplicationState appState = [RCTSharedApplication() applicationState];
  BOOL isRunningInFullScreen = CGRectEqualToRect([UIApplication sharedApplication].delegate.window.frame, [UIApplication sharedApplication].delegate.window.screen.bounds);

  // Update when we go from portrait to landscape, or landscape to portrait
  if ((((UIInterfaceOrientationIsPortrait(_currentInterfaceOrientation) &&
       !UIInterfaceOrientationIsPortrait(nextOrientation)) ||
      (UIInterfaceOrientationIsLandscape(_currentInterfaceOrientation) &&
       !UIInterfaceOrientationIsLandscape(nextOrientation)) || (isRunningInFullScreen != _isFullscreen || !isRunningInFullScreen)))
      && appState == UIApplicationStateActive) {
      #pragma clang diagnostic push
      #pragma clang diagnostic ignored "-Wdeprecated-declarations"
          [[_moduleRegistry moduleForName:"EventDispatcher"] sendDeviceEventWithName:@"didUpdateDimensions"
                                                                                body:RCTExportedDimensions(_moduleRegistry)];
            _currentInterfaceOrientation = nextOrientation;
            _isFullscreen = isRunningInFullScreen;
      #pragma clang diagnostic pop
  }


}
```

b) _interfaceFrameDidChange

Here we make sure we run it when the app is in active state.

```
- (void)_interfaceFrameDidChange
{
  NSDictionary *nextInterfaceDimensions = RCTExportedDimensions(_moduleRegistry);
    UIApplicationState appState = [RCTSharedApplication() applicationState];

  if (!([nextInterfaceDimensions isEqual:_currentInterfaceDimensions]) && appState == UIApplicationStateActive) {
    #pragma clang diagnostic push
    #pragma clang diagnostic ignored "-Wdeprecated-declarations"
        [[_moduleRegistry moduleForName:"EventDispatcher"] sendDeviceEventWithName:@"didUpdateDimensions"
                                                                              body:nextInterfaceDimensions];
          _currentInterfaceDimensions = nextInterfaceDimensions;
    #pragma clang diagnostic pop
  }
}
```

c) Add the \_isFullscreen variable to hold the previous state for multitasking view.

```
@implementation RCTDeviceInfo {
  UIInterfaceOrientation _currentInterfaceOrientation;
  NSDictionary *_currentInterfaceDimensions;
  BOOL _isFullscreen;
}
```

d) We need to make sure to run the interfaceOrientationDidChange instead of interfaceFrameDidChange when the UIApplicationDidBecomeActiveNotification occurs.

```
  [[NSNotificationCenter defaultCenter] addObserver:self
                                           selector:@selector(interfaceOrientationDidChange)
                                               name:UIApplicationDidBecomeActiveNotification
                                             object:nil];
```
[CATEGORY] [TYPE] - Message

## Test Plan

iPad mini tests:
Test 1:
1. Launch app in portrait, open chat - no sidebar visible.
2. Switch to landscape - sidebar shows.
3. Put app to background.
4. Put app back to foreground - make sure the side menu doesn't flicker.

Test 2:
1. Launch app in portrait, open chat - no sidebar visible.
2. Switch to landscape - sidebar shows.
3. Put app to background. Switch orientation back to portrait.
4. Put app back to foreground - make sure the side menu hides again as it should be in portrait.

iPad Pro:

iPad mini scenario 1 applies.

Scenario 2 does not as the screen is too wide in both orientations and ipad pro shows sidebar always.

Scenario 1:

1.  launch the app.
2. Make sure you're in landscape mode.
3. See split screen with some other app. Make sure the side bar is visible.
4. Play with the size of the view, resize it a bit. When the view shrinks it should hide the sidebar, when it grows it should show it.
5. Move the app to background and back to foreground, please observe there are no flickers.

Scenario 2:

1.  Launch the app.
2. Make sure you're in landscape mode.
3. Make the multitasking view and make Expensify app a slide over app.
4. Move back to fullscreen/split screen. Make sure the menu is shown accordingly
5. Move the app to background and back to foreground, please observe there are no flickers.

iPhones:

Please perform standard smoke tests on transformation changes.

parasharrajat commented:
I see that PR can be improved in the following ways. 

- Formatting of OP.
- It would be good to link existing RN issues to the PR. 
- Videos and reproduction.

lbaldy commented:
@parasharrajat I added the videos of me reproducing the issue and testing this with the fix applied. Had to upload to youtube due to size constrains here on Github. 

- Formatting of OP.
I updated the formatting to make the PR description a little bit more legible. If you see anything else that should be improved please let me know.
- It would be good to link existing RN issues to the PR.
I am not sure which issues you are referring to? I linked the original issue from the Expensify/App repo already.
- Videos and reproduction.
Added.

Please let me know if you believe anything else has to be imoproved/corrected. 

Thanks.

parasharrajat commented:
I will test this today.

lbaldy commented:
Hi @parasharrajat  any news on this one?

parasharrajat commented:
On my list. Just me give a day more. Sorry for the delay. I broke my Mac past week so things got piled up.

parasharrajat commented:
> It would be good to link existing RN issues to the PR.

Existing issues about this bug we are trying to solve. Please search RN codebase and link any whatever looks the same. In case there are none, create a new one which we can refer to while submitting the PR to RN.

lbaldy commented:
@parasharrajat  I updated the original post. 

I added the https://github.com/facebook/react-native/issues/29290 which was mentioned in the original https://github.com/Expensify/App/issues/2727.

parasharrajat commented:
I will start reviewing this in few minutes.

parasharrajat reviewed (changes requested):
I see your test cases. But you didn't mention test process. How can a person test these code changes on E/App? Please add this details.

parasharrajat commented:
> We need to make sure to run the interfaceOrientationDidChange instead of interfaceFrameDidChange when the UIApplicationDidBecomeActiveNotification occurs.

Why?

lbaldy commented:
> > We need to make sure to run the interfaceOrientationDidChange instead of interfaceFrameDidChange when the UIApplicationDidBecomeActiveNotification occurs.
> 
> Why?

So that in case of this event we also have the extended handler, which considers the fullscreen as well as active state.

lbaldy commented on React/CoreModules/RCTDeviceInfo.mm:
```diff
@@ -174,21 +175,25 @@ - (void)interfaceOrientationDidChange
-        sendDeviceEventWithName:@"didUpdateDimensions"
-                           body:RCTExportedDimensions(_moduleRegistry, _bridge)];
-#pragma clang diagnostic pop
+       !UIInterfaceOrientationIsLandscape(nextOrientation)) || (isRunningInFullScreen != _isFullscreen || !isRunningInFullScreen)))
```
done, in https://github.com/Expensify/react-native/pull/11/commits/b0c3bc71c3a60f250a193b13d6da9b9f11c168e0

lbaldy commented on React/CoreModules/RCTDeviceInfo.mm:
```diff
@@ -202,16 +207,16 @@ - (void)interfaceFrameDidChange
+          _currentInterfaceDimensions = nextInterfaceDimensions;
+    #pragma clang diagnostic pop
   }
-
-  _currentInterfaceDimensions = nextInterfaceDimensions;
```
I added a comment in code, but that's moved inside so that we make sure that the state between JS and Obj-C is synced. Meaning we only update the 'Native' state when it's published to JS.

lbaldy commented:
Updated

parasharrajat commented:
Interesting, I am trying to test the iOS build in landscape mode. I commented the code that was used to prevent the landscape but still app is not going to landscape. This is giving me nuts. 

Any guesses?

parasharrajat commented on React/CoreModules/RCTDeviceInfo.mm:
```diff
@@ -202,16 +213,20 @@ - (void)interfaceFrameDidChange
+    #pragma clang diagnostic ignored "-Wdeprecated-declarations"
+        [[_moduleRegistry moduleForName:"EventDispatcher"] sendDeviceEventWithName:@"didUpdateDimensions"
+                                                                              body:nextInterfaceDimensions];
+          // We only want to track the current _currentInterfaceOrientation and _isFullscreen only 
+          // when it happens and only when it is published, thus it has to happen inside the if condition.
```
```suggestion
          // We only want to track the current _currentInterfaceOrientation and _isFullscreen only 
          // when it happens and only when it is published.
```

instead of it happens, we should tell what exactly is happening? This is not clear to me.

parasharrajat commented on React/CoreModules/RCTDeviceInfo.mm:
```diff
@@ -174,21 +175,31 @@ - (void)interfaceOrientationDidChange
+      #pragma clang diagnostic ignored "-Wdeprecated-declarations"
+          [[_moduleRegistry moduleForName:"EventDispatcher"] sendDeviceEventWithName:@"didUpdateDimensions"
+                                                                                body:RCTExportedDimensions(_moduleRegistry, _bridge)];
+            // We only want to track the current _currentInterfaceOrientation and _isFullscreen only 
+            // when it happens and only when it is published, thus it has to happen inside the if condition.
```
Same here.

parasharrajat commented on React/CoreModules/RCTDeviceInfo.mm:
```diff
@@ -174,21 +175,31 @@ - (void)interfaceOrientationDidChange
-        sendDeviceEventWithName:@"didUpdateDimensions"
-                           body:RCTExportedDimensions(_moduleRegistry, _bridge)];
-#pragma clang diagnostic pop
+  // also update when the fullscreen statea changes (multitasking) and only when the app is in active state
```
```suggestion
  // Also update when the fullscreen state changes (multitasking) and only when the app is in active state.
```

parasharrajat reviewed (changes requested):
I am running final tests.

parasharrajat reviewed (commented):
LGTM. 

cc: @iwiznia Please review this PR.

sketchydroide commented on React/CoreModules/RCTDeviceInfo.mm:
```diff
@@ -174,21 +175,31 @@ - (void)interfaceOrientationDidChange
+    
+  
+  BOOL isActive = appState == UIApplicationStateActive;
+  BOOL isChangingFullscreen = (isRunningInFullScreen != _isFullscreen || !isRunningInFullScreen);
```
this is a weird condition, and the name does not seem to match the condition
If we are matining the condition, we need a comment to explainf why this is not simply 
```BOOL isChangingFullscreen = isRunningInFullScreen != _isFullscreen;```
The condition could be simplified to
```BOOL isChangingFullscreen = !isRunningInFullScreen || !_isFullscreen;```
And again I'm not entirely sure that condition makes sense

lbaldy replied:
This condition indeed can be shortened to what you suggested, good catch. Thanks.

The condition is needed and called this way because it catches two scenarios:

a) Simple resize when the app isn't in the fullscreen mode right now - the !isRunningInFullScreen
b) Resize when the app gets from slide over mode to fullscreen - isRunningInFullScreen != _isFullscreen

Anyway, I commented this line with the above.

sketchydroide reviewed (changes requested):
I have a question about one of the conditions about `isChangingFullscreen`

The rest seems good

lbaldy commented:
Updated

parasharrajat reviewed (approved):
LGTM

:ribbon: :eyes: :ribbon:  C+ reviewed

Justicea83 reviewed (approved):
LGTM!

sketchydroide commented:
@lbaldy I think @iwiznia is still reviwieng this PR, he asked me to take a look as I'm more confortable with ObjC, but I think his input is still important

parasharrajat commented:
> shall we simply wait for inputs from @iwiznia.

This one.
